R/http.R

```diff
@@ -0,0 +1,106 @@
+#' @title Travis CI HTTP Requests
+#'
+#' @description This is the workhorse function for executing API requests for
+#'   Travis CI.
+#'
+#' @import httr
+#' @importFrom jsonlite fromJSON
+#'
+#' @details This is mostly an internal function for executing API requests. In
+#'   almost all cases, users do not need to access this directly.
+#'
+#' @param verb A character string containing an HTTP verb, defaulting to `GET`.
+#' @param path A character string with the API endpoint (should begin with a
+#'   slash).
+#' @param query A list specifying any query string arguments to pass to the API.
+#'   This is used to pass the API token.
+#' @param body A named list or character string of what should be passed in the
+#'   request. Corresponds to the "-d" argument of the `curl` command.
+#' @template endpoint
+#' @param encode Encoding format. See [httr::POST].
+#'
+#' @return The JSON response, or the relevant error.
+#'
+#' @export
+travisHTTP <- function(verb = "GET",
+                       path = "",
+                       query = list(),
+                       body = "",
+                       endpoint = ".org",
+                       encode = "json") {
+  url <- sprintf("https://api.travis-ci%s%s", endpoint, path)
+
+  auth_travis()
+  api_token <- read_token()
+  # set user agent
+  ua <- user_agent("http://github.com/ropenscilabs/travis")
+
+  if (verb == "GET") {
+    resp <- GET(url,
+                add_headers(Authorization = sprintf("token %s", api_token),
+                            "Travis-API-Version" = 3),
+                query = query, encode = encode, ua, accept_json(),
+                content_type_json()
+    )
+  } else if (verb == "DELETE") {
+    resp <- DELETE(url,
+                   add_headers(Authorization = sprintf("token %s", api_token),
+                               "Travis-API-Version" = 3),
+                   query = query, encode = encode, ua, accept_json(),
+                   content_type_json()
+    )
+  } else if (verb == "POST") {
+    resp <- POST(url,
+                 add_headers(Authorization = sprintf("token %s", api_token),
+                             "Travis-API-Version" = 3),
+                 body = body, query = query, encode = encode, ua,
+                 accept_json()
+    )
+  }
+
+  # for travis_delete_var()
+  if (http_type(resp) == "application/octet-stream") {
+    return(resp)
+  }
+  if (http_type(resp) == "text/plain") {
+    return(resp)
+  }
+
+  parsed <- fromJSON(content(resp, "text", encoding = "UTF-8"), simplifyVector = FALSE)
+
+  # catch specific errors
+  if (any(grepl("error_type", names(parsed))) && parsed$error_type == "job_already_running") {
```

Comment:
Do these errors come with a specific HTTP response code? Can we check `is.null(parsed$error_type)` beforehand, maybe extract this special handling into a function?

Reply:
Unfortunately not, that's why I went this way. I knew this would lead to a discussion but I see no easier way right now.

A function would be a bit more clean probably, yes.

Reply:
Objects of this type look as follows

<img width="598" alt="Screenshot 2019-11-10 at 20 37 32" src="https://user-images.githubusercontent.com/8430564/68549612-705be780-03fa-11ea-92fe-6379d84156e0.png">

Outsourced the special error catching into a helper fun 👍